govuk-one-login/account-interventions-service
# ATB-1137 | TxMA Egress Messages

## Proposed changes
- added check for stale events
- added send audit event function
- refactored account state engine class to only return new account state (not partial db update command)
### What changed
- stale event check: when a valid event is received, the processor lambda checks that its timestamp value is in the future wrt the sentAt (or appliedAt) values for the user as retrieved from the DB. 
- add sentAt field in update command and set to equal timestamp in ms received in the event
- moved the call to `buildPartialUpdateAccountStateCommand` from the Account State Engine class to the processor
- added calls to the send audit event function to implement notifications to TxMA as per linked ticket

### Why did it change
To send notifications of outcome of processed event

### Issue tracking

- [ATB-1137](https://govukverify.atlassian.net/browse/ATB-1137)

## Testing
Deployed to AWS account:

timestamp logic
- send valid event timestamped at T, send valid event timestamped at T -1 --> second event is rejected ✅ 
- send valid intervention event: check that appliedAt, sentAt, updatedAt fields are populated with appropriate values ✅ 
<img width="508" alt="Screenshot 2023-11-08 at 17 26 44" src="https://github.com/govuk-one-login/account-interventions-service/assets/110468890/1588e6a1-188e-4730-b49b-44797002d515">

- send valid user action event: check that resetPasswordAt/resetIdentityAt, updatedAt fields are populated with appropriate values ✅ 
<img width="611" alt="Screenshot 2023-11-08 at 17 32 53" src="https://github.com/govuk-one-login/account-interventions-service/assets/110468890/147a66f6-b186-4d87-9292-8ebd363b217d">

txma publishing
- `AIS_INTERVENTION_TRANSITION_APPLIED`
<img width="611" alt="Screenshot 2023-11-08 at 17 34 07" src="https://github.com/govuk-one-login/account-interventions-service/assets/110468890/81418573-54c5-4d2c-9e51-298d2853381f">

- `AIS_INTERVENTION_TRANSITION_IGNORED`
<img width="611" alt="Screenshot 2023-11-08 at 17 34 57" src="https://github.com/govuk-one-login/account-interventions-service/assets/110468890/698ba64b-6e5a-454d-912d-7cc4022e9668">

- `AIS_INTERVENTION_IGNORED_STALE`
<img width="611" alt="Screenshot 2023-11-08 at 17 34 28" src="https://github.com/govuk-one-login/account-interventions-service/assets/110468890/281e3165-834e-4391-b77f-82355a6d78e1">

- `AIS_INTERVENTION_IGNORED_INFUTURE` (message is re added to the queue)
<img width="611" alt="Screenshot 2023-11-08 at 17 36 34" src="https://github.com/govuk-one-login/account-interventions-service/assets/110468890/aee13061-c064-429f-90f7-cd5fdec79d24">

- `AIS_INTERVENTION_IGNORED_ACCOUNT_DELETED`
<img width="638" alt="Screenshot 2023-11-08 at 17 41 39" src="https://github.com/govuk-one-login/account-interventions-service/assets/110468890/fd36fb77-85cf-4a3b-9484-b6efe9b614da">


## Checklists
- [x] Did not commit any not-required changes to the `src/infra/**/samconfig.toml`
- [x] Tested changes and included test evidence in the PR, if appropriate
- [x] Included all required tags and other properties for any new resources in the SAM template
- [x] Ensured that any new resources in the SAM Template follow appropriate naming conventions
- [x] Ensured that naming of new resources is compatible with deploying multiple stacks with custom stack names during development
- [x] Ensured that no log lines include PII or other sensitive data
- [x] Implemented unit testing for any new logic implemented, if appropriate
- [x] Ensured that all commits in this PR are signed
- [x] Ensured appropriate code coverage is maintained by unit tests


[ATB-1137]: https://govukverify.atlassian.net/browse/ATB-1137?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
